File: go/patterns/tmp2/solution/twosum.go
package solution

import "errors"

type TwoSum struct {
	Nums   []int
	Target int
}

func (s *TwoSum) Find() ([]int, error) {
	if len(s.Nums) < 2 {
		return nil, errors.New("Invalid count of nums")
	}

	seen := make(map[int]int) // int to idx for complement

	for idx, num := range s.Nums {
		complement := s.Target - num
		if compIdx, found := seen[complement]; found {
			return []int{compIdx, idx}, nil
		}
		seen[num] = idx
	}
	return nil, errors.New("Invalid indices from nums")
}

File: typescript/tmp2/two_sum.ts

function run(nums: number[], target: number): number[] {
    const seen = new Map<number, number>();

    for (let i = 0; i < nums.length; i++) {
        const complement =  target - nums[i];
        if (seen.has(complement)) {
            return [seen.get(complement)!, i];
        }
        seen.set(nums[i], i);
    }
    return [];
}

function main() {
    const nums = [1, 2, 3, 4];
    const target = 7;
    const result = run(nums, target);
    console.log(result);
}

main();
